src/main/java/de/azapps/kafkabackup/sink/BackupSinkTask.java

```diff
@@ -50,15 +71,49 @@ public void start(Map<String, String> props, OffsetSink overrideOffsetSink) {
             if(overrideOffsetSink != null) {
                 offsetSink = overrideOffsetSink;
             } else {
-                AdminClient adminClient = AdminClient.create(config.adminConfig());
+                adminClient = AdminClient.create(config.adminConfig());
                 offsetSink = new OffsetSink(adminClient, targetDir);
             }
+
+            if (overrideEndOffsetReader != null) {
+                this.endOffsetReader = overrideEndOffsetReader;
+            } else {
+                endOffsetReader = new EndOffsetReader(bootstrapServers);
+            }
+
+            if (overrideExitFunction != null) {
+                this.exitFunction = overrideExitFunction;
+            } else {
+                this.exitFunction = System::exit;
+            }
+
             log.debug("Initialized BackupSinkTask with target dir {}", targetDir);
         } catch (IOException e) {
             throw new RuntimeException(e);
         }
     }
 
+    /**
+     * Check for end-offsets. Terminate if all offsets >= end-offsets
+     */
+    private void terminateIfCompleted() {
+        boolean terminate = true;
+        for (TopicPartition partition : endOffsets.keySet()) {
+            Long endOffset = endOffsets.get(partition);
+            Long currentOffset = currentOffsets.getOrDefault(partition, -1L);
+
+            if (currentOffset < endOffset - 1) {
+                terminate = false;
+                break;
```

Comment:
I think you could even `return` here – or do you mislike doing so?

Reply:
I'm a Scala developer, 'return' is not common there :-) . Will fix.